### PR TITLE
Driver `/solve` functionality part 1

### PR DESCRIPTION
**Pull request by sistemd**
Progress on https://github.com/cowprotocol/services/issues/895. Implements part of the design outlined in https://github.com/cowprotocol/services/issues/897.

Implements an interface for talking to the external solvers. Sketches out a skeleton for the core logic and some of the externals.

### Test Plan

Will write API-level tests later down the line.

### CODE REVIEW

File: crates/driver/Cargo.toml [hunk 1]
```diff
@@ -15,26 +15,34 @@ name = "driver"
 path = "src/main.rs"
 
 [dependencies]
+num = "0.4"
```
Comment (nlordell) [inline crates/driver/Cargo.toml:18]: I noticed at first that these don't use `{ workspace = true }` dependencies (which I personally like).

At first, I was going to suggest to change them back, but I realize that this is probably intentional as we intend to move it into a separate repo. Is this correct?

File: crates/driver/src/solver/dto/auction.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    tokens: BTreeMap<H160, Token>,
+    orders: BTreeMap<usize, Order>,
+    amms: BTreeMap<H160, Amm>,
+    metadata: Option<Metadata>,
+}
+
+#[derive(Debug, Serialize)]
+struct Order {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    id: Option<String>,
```
Comment (nlordell) [inline crates/driver/src/solver/dto/auction.rs:29]: All orders should have a UID (note that this is different than the `LimitOrder::id` field that we used to have in the `solver` - as this represents a protocol order which has an actual mandatory and unique ID).

File: crates/driver/src/solver/mod.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    pub async fn solve(&self, auction: Auction) -> Result<Solution, Error> {
+        // TODO Ask about all the `config` stuff in DefaultHttpSolverApi, what is every
+        // field for exactly?
+        let mut url = self.url.join("solve").unwrap();
+        let time_limit = auction.deadline.solver_time_limit()?;
+        url.query_pairs_mut()
+            .append_pair("auction_id", &auction.id.0.to_string())
+            .append_pair("instance_name", &self.instance_name(auction.id))
```
Comment (nlordell) [inline crates/driver/src/solver/mod.rs:59]: For example, this is literally telling the solver what to name the instance file. I think you could get it do crazy things by specifying a path (`../../etc/passwd` 😛).

File: crates/driver/src/util/serialize.rs [hunk 4]
```diff
@@ -0,0 +1,30 @@
+use serde::{Deserialize, Serialize};
+
+/// Serializes [`ethereum_types::U256`] as a decimal string.
```
Comment (nlordell) [inline crates/driver/src/util/serialize.rs:3]: I saw these change to `serde_as` in the other PR (which I think is more correct).

**Review by nlordell — APPROVED** [review]
Code looks good.

I think there are a few assumptions related to https://github.com/cowprotocol/services/issues/895#issuecomment-1337498190 in this PR as well. I already brought this up in #910, and I think this PR is good WRT driver <> solver communication (which is its focus anyway).

File: crates/driver/src/solver/dto/solution.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    Coordinates(Coordinates),
+    #[serde(deserialize_with = "execution_plan_internal")]
+    Internal,
+}
+
+#[derive(Debug, Deserialize)]
+struct Coordinates {
+    sequence: u32,
+    position: u32,
+    #[serde(default)]
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/driver/src/solver/dto/solution.rs:138]: Nice catch!

File: crates/driver/src/solver/mod.rs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            "application/json".parse().unwrap(),
+        );
+        headers.insert(reqwest::header::ACCEPT, "application/json".parse().unwrap());
+        // TODO(#907) Also add an auth header
+        Self {
+            url,
+            client: reqwest::ClientBuilder::new()
+                .default_headers(headers)
+                .build()
+                .unwrap(),
```
Comment (MartinquaXD) [inline crates/driver/src/solver/mod.rs:46]: The client should probably passed into the `Solver` to be able to configure things like timeouts at a centralized place.

**Pull request merged**